### DISCUSSION TITLE
Customize Langchain callback handler to add spans to existing traces created by separate thread/process

### DISCUSSION THREAD

**Discussion by Chengdyc** [langfuse/langfuse#2351] [Support]
Hi, we have a use case where we have a workflow that is orchestrated with a workflow engine. Each step in the workflow would invoke Langchain to perform some work. I simple example would be one step to perform a vector store retrieval, and another to call LLM to generate the completion. (This toy example doesn't warrant the need for workflow engine but the real-world examples do).

Since each step of the workflow could be executed in a different thread/process, the default behavior of Langchain callback handler is to create a separate trace/run-id for each step. This results in multiple traces in Langfuse. I tried overriding the run-id but that just overwrote the trace to be whatever the last step in the workflow is. ideally what we want is to have a top-level trace that correspond to the workflow, then each step in the workflow would create a span in the trace.

Is this something we can achieve via Langfuse SDK currently?

**Comment by GMartin-dev** [comment]
Hi there! I think you could use this:
https://github.com/orgs/langfuse/discussions/2172#discussioncomment-9769762

**Reply by Chengdyc** [reply]
Thanks! I tried that and it works mostly, but I'm having problems with 2 issues:
1. the output of the top-level trace gets overwritten occasionally by the output from the last step of my workflow. It seems there's a race condition here. 
2. one of the steps in the workflow makes a tool call, and Langfuse raises an exception complaining that parent run is not found.

for more context, my workflow has 4 steps:
1. agent call to LLM to pick a tool (but not execute the tool)
2. execute the tool chosen by LLM (or exit if no tool was picked). in this case the tool is vectorstore retriever
3. generate completion using the context fetched by the tool
4. call LLM to check for hallucination, return completion if there's no hallucination.

The code is structured this way:
1. one process to start the workflow. logically, the code looks like this:
```
langfuse_context.configure(...)

@observe()
async def _start_workflow(input):
   trace_id = langfuse_context.get_current_trace_id()

   handle = workflow_engine.start_workflow(trace_id, input)   # trace_id will be used to override the run_id in each step of the workflow.
   output = await handle

   langfuse_context.update_current_trace(name="my_workflow")
   langfuse_context.update_current_observation(
      input=input, output=output
   )

await _start_workflow(input={ "question": "<user question>" })

```

2. The steps of the workflow run in a separate process (or potentially each in different processes in a distributed environment), e.g.

```
def step1(trace_id, input):
   chain.invoke(
      input,
      config={ 
         "run_id"=trace_id,
         "callbacks"=[ handler ]
      }
   )
```

how do I mitigate the issue where the top-level trace's output gets overwritten? thanks.

**Comment by GMartin-dev** [comment]
I know this approach is kinda a hacky, It does not work for all scenarios like with the batch method.
Also I'm seeing some issues on some runnable events missing the parent, more than overriding of the top trace i'm seeing that it duplicates a span with the run name, it does not bother me that much, but my approach is slightly diff.
I do not use @observe() and I initialize the handler with a manual run name and I only make sure the initial run_id is the same and with just that it kinda works for me.
Raw and quick example on how I do it.

```Python
def get_langfuse_handler(run_name: str):
    """We make sure we have a trace from initial handler instance."""
    langfuse = Langfuse()
    trace = langfuse.trace(name=run_name)
    return CallbackHandler(trace_name=run_name, stateful_client=trace)

#This is the key... this id must be the same across processes.
run_id = uuid.uuid4()

#THe handler can be reused or totally new depending of the process.
handler = get_langfuse_handler("Name to wrap the whole sequence of stuff.")
callbacks = [handler]

runnable_object1.invoke(
        input=kwargs,
        config={"callbacks": callbacks, "run_id": self.run_id},
    )
span_client = handler.trace.span(
        trace_id=self.trace_handler.trace.id,
        name="Custom process tracing here",
        metadata={},
        input=query,
        version=version_label,
    )
process_output = custom_process(data)           
span_client.end(
            output=process_output,
            name="Override custom process tracing name",
        )
runnable_object2.invoke(
        input=kwargs,
        config={"callbacks": callbacks, "run_id": self.run_id},
    )
            
    ```
Each invoke or function can be executed in diff processes, like diff celery tasks and they log all together for me.

**Reply by Chengdyc** [reply]
I switched to using langfuse.trace(), but the issue still persists. seems there's still a race condition where the output of the top-level trace get overwritten by the last step in the workflow.

```
langfuse_client = Langfuse()

async def _start_workflow(input):
   trace = langfuse_client.trace(name="my_workflow")
   trace_id = trace.id

   handle = workflow_engine.start_workflow(trace_id, input)   # trace_id will be used to override the run_id in each step of the workflow.
   output = await handle

   trace.update(
      name="my_workflow",
      metadata={"workflow_id": workflow_id},
      input={"question": question},
      output=answer,
   )

await _start_workflow(input={ "question": "<user question>" })

langfuse_client.fluse()
```

I fixed the 'parent run is not found' issue by wrapping the tool calls in a RunnableLambda that calls the invoke() method directly

```
def call_invoke(input: Dict[str, Any], config: RunnableConfig):
    return selected_tool.invoke(input, config)

wrapper = RunnableLambda(
    call_invoke
).with_config({"run_name": "invoke_tools"})
```

**Reply by GMartin-dev** [reply]
Did you tried adding a manual "flush" at the end of each intermediary step?

**Reply by Chengdyc** [reply]
no. all the intermediary steps use the callback handlers, they don't have access to langfuse client directly. is there a way to flush using the callback handlers? or we'd have to use the langfuse client?

also, we're using langserve to expose the workflow as an API. so the top-level trace is created by langserve itself. ideally the trace should be structured like this:
```
- trace top level trace
   - span /my-api-path
      - span intermediate step 1
      - span intermediate step 2
      - span intermediate step 3
```

but currently using the method described earlier. the intermediate steps are attached to the top level trace

```
- trace top level trace
   - span /my-api-path
   - span intermediate step 1
   - span intermediate step 2
   - span intermediate step 3
```

is there a way to attach the intermediate steps to the span that langserve creates?

**Reply by GMartin-dev** [reply]
Yes you can call flush from the handler directly:
https://github.com/langfuse/langfuse-python/blob/4a1be9b3355b9e5dad853bd8f2ce08a0c8f72b52/langfuse/utils/base_callback_handler.py#L124

I'm not using langserve so I can not help you much there, perhaps that introduces some additional nuances.

**Reply by Chengdyc** [reply]
Thanks, setting flush_at=1 in the callback handler seems to do the trick.

**Reply by GMartin-dev** [reply]
AH I did not know that trick! Happy it finally worked!
